---
# SPDX-FileCopyrightText: 2020 Sobeston and zig.guide contributors
# SPDX-FileCopyrightText: 2024 BratishkaErik and zigtutor.ru contributors
#
# SPDX-License-Identifier: MIT

title: Errors

slug: language-basics/errors
sidebar:
    order: 8
---

import { Code } from "@astrojs/starlight/components";

import ErrorSet from "08.error-set.zig?raw";
import ErrorSetCoercion from "08.error-set-coercion.zig?raw";
import ErrorUnion from "08.error-union.zig?raw";
import ErrorPayloadCapture from "08.error-payload-capture.zig?raw";
import ErrorTry from "08.error-try.zig?raw";
import Errdefer from "08.errdefer.zig?raw";
import ErrorSetInferred from "08.error-set-inferred.zig?raw";
import ErrorSetMerge from "08.error-set-merge.zig?raw";

An error set is like an enum (details on Zig's enums later), where each error in
the set is a value. There are no exceptions in Zig; errors are values. Let's
create an error set.

<Code code={ErrorSet} lang="zig"/>

Error sets coerce to their supersets.

<Code code={ErrorSetCoercion} collapse="1-8" lang="zig"/>

An error set type and another type can be combined with the `!` operator to form
an error union type. Values of these types may be an error value or a value of
the other type.

Let's create a value of an error union type. Here
[`catch`](https://ziglang.org/documentation/master/#catch) is used, which is
followed by an expression which is evaluated when the value preceding it is an
error. The catch here is used to provide a fallback value, but could instead be
a [`noreturn`](https://ziglang.org/documentation/master/#noreturn) - the type of
`return`, `while (true)` and others.

<Code code={ErrorUnion} collapse="1-10" lang="zig"/>

Functions often return error unions. Here's one using a catch, where the `|err|`
syntax receives the value of the error. This is called **payload capturing**,
and is used similarly in many places. We'll talk about it in more detail later
in the chapter. Side note: some languages use similar syntax for lambdas - this
is not true for Zig.

<Code code={ErrorPayloadCapture} collapse="1-2" lang="zig"/>

`try x` is a shortcut for `x catch |err| return err`, and is commonly used where
handling an error isn't appropriate. Zig's
[`try`](https://ziglang.org/documentation/master/#try) and
[`catch`](https://ziglang.org/documentation/master/#catch) are unrelated to
try-catch in other languages.

<Code code={ErrorTry} collapse="1-6" lang="zig"/>

[`errdefer`](https://ziglang.org/documentation/master/#errdefer) works like
[`defer`](https://ziglang.org/documentation/master/#defer), but only executing
when the function is returned from with an error inside of the
[`errdefer`](https://ziglang.org/documentation/master/#errdefer)'s block.

<Code code={Errdefer} collapse="1-6" lang="zig"/>

Error unions returned from a function can have their error sets inferred by not
having an explicit error set. This inferred error set contains all possible
errors that the function may return.

<Code code={ErrorSetInferred} lang="zig"/>

Error sets can be merged.

<Code code={ErrorSetMerge} lang="zig"/>

`anyerror` is the global error set, which due to being the superset of all error
sets, can have an error from any set coerced to it. Its usage should be
generally avoided.
